### ISSUE TITLE
Bug detection: enter valgrind

### ISSUE THREAD

**Issue by catull** [att/ast#271]
# meson script support

Unfortunately, meson does not support this:
     `meson test -v -V`

`-V` would run the test scripts under `valgrind`, collecting a wealth of information, see below.

# shtests: main test script runner

Deep down the main shell test driver is a script called shtests.
It is located in src/cmd/ksh93/tests/shtests.

It accepts some parameters, only one of them is currently supported in the meson script: -v.

Being an explorer who goes to the bottom of things, I dug out this way to wake up more options to shtests.

# Pre-conditions
 - valgrind must be available for your OS (good for Linux, macOS only up to 10.12)
 - must explicitly set SHELL to the built ksh

# Steps to run valgrind-ed tests

So this is how I did it under Fedora:
 - `dnf install valgrind`
 - after `ninja -C build`; cd build
 - SHELL=`pwd`/src/cmd/ksh93/ksh
 - cd ../src/cmd/ksh93/tests
 - $SHELL ./shtests -v -V

# Impact

It displays a lot of info.
Of course, this slows down the tests, I would say they are more than 10x - 20x slower, rough guess.

# Positive run

If the test runs without any valgrind issues, it displays the same lines as with `./shtests -v`:
```log
test alias begins at 2017-12-27+19:11:58
test alias passed at 2017-12-27+19:12:03 [ 23 tests 0 errors ]
test alias(C.UTF-8) begins at 2017-12-27+19:12:03
test alias(C.UTF-8) passed at 2017-12-27+19:12:08 [ 23 tests 0 errors ]
test alias(shcomp) begins at 2017-12-27+19:12:08
test alias(shcomp) passed at 2017-12-27+19:12:13 [ 23 tests 0 errors ]
```
# Sort of negative run

Here's an example for a chatty run, on `append`:

```log
test append begins at 2017-12-27+19:12:13

	Invalid read of size 2
	    build/../src/cmd/ksh93/sh/nvtype.c:1361:nv_type
	    build/../src/cmd/ksh93/sh/name.c:612:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main
	Address 0x5f2f898 is 24 bytes inside a block of size 66 free'd
	    /builddir/build/BUILD/valgrind-3.13.0/coregrind/m_replacemalloc/vg_replace_malloc.c:530:free
	    build/../src/cmd/ksh93/sh/name.c:1387:nv_delete
	    build/../src/cmd/ksh93/sh/array.c:592:array_putval
	    build/../src/cmd/ksh93/sh/nvdisc.c:152:nv_putv
	    build/../src/cmd/ksh93/sh/name.c:2658:_nv_unset
	    build/../src/cmd/ksh93/sh/name.c:570:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main
	Block was alloc'd at
	    /builddir/build/BUILD/valgrind-3.13.0/coregrind/m_replacemalloc/vg_replace_malloc.c:711:calloc
	    build/../src/cmd/ksh93/sh/nvdisc.c:840:newnode
	    build/../src/cmd/ksh93/sh/nvdisc.c:1097:nv_search
	    build/../src/cmd/ksh93/sh/name.c:1227:nv_create
	    build/../src/cmd/ksh93/sh/name.c:1532:nv_open
	    build/../src/cmd/ksh93/sh/name.c:373:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main

	Invalid read of size 8
	    build/../src/cmd/ksh93/sh/nvtype.c:1366:nv_type
	    build/../src/cmd/ksh93/sh/name.c:612:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main
	Address 0x5f2f8a0 is 32 bytes inside a block of size 66 free'd
	    /builddir/build/BUILD/valgrind-3.13.0/coregrind/m_replacemalloc/vg_replace_malloc.c:530:free
	    build/../src/cmd/ksh93/sh/name.c:1387:nv_delete
	    build/../src/cmd/ksh93/sh/array.c:592:array_putval
	    build/../src/cmd/ksh93/sh/nvdisc.c:152:nv_putv
	    build/../src/cmd/ksh93/sh/name.c:2658:_nv_unset
	    build/../src/cmd/ksh93/sh/name.c:570:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main
	Block was alloc'd at
	    /builddir/build/BUILD/valgrind-3.13.0/coregrind/m_replacemalloc/vg_replace_malloc.c:711:calloc
	    build/../src/cmd/ksh93/sh/nvdisc.c:840:newnode
	    build/../src/cmd/ksh93/sh/nvdisc.c:1097:nv_search
	    build/../src/cmd/ksh93/sh/name.c:1227:nv_create
	    build/../src/cmd/ksh93/sh/name.c:1532:nv_open
	    build/../src/cmd/ksh93/sh/name.c:373:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main

	Conditional jump or move depends on uninitialised value(s)
	    /builddir/build/BUILD/valgrind-3.13.0/memcheck/../shared/vg_replace_strmem.c:1099:__memcmp_sse4_1
	    build/../src/cmd/ksh93/sh/name.c:217:nv_addnode
	    build/../src/cmd/ksh93/sh/name.c:1593:nv_open
	    build/../src/cmd/ksh93/sh/name.c:689:sh_setlist
	    build/../src/cmd/ksh93/sh/xec.c:1148:sh_exec
	    build/../src/cmd/ksh93/sh/main.c:609:exfile
	    build/../src/cmd/ksh93/sh/main.c:378:sh_main
	    build/../src/cmd/ksh93/sh/pmain.c:43:main

test append failed at 2017-12-27+19:12:17 with exit code 3 [ 19 tests 3 errors ]
```

# Conclusion

If we run the tests under `valgrind`, many more will fail.
E.g. the test `append` - depicted above, has never failed for me, ever.
With `valgrind` turned on, it does fail.

So this means switching on `valgrind` is only meaningful when the code base has become much more stable and manageable.

In other words, now is not a good time to incorporated it.

I do hope though that it helps me spot the SIGABRT and SIGSEV under macOS 10.12.6.

It would be great if the meson script supported `valgrind` directly.

**Comment by krader1961** [comment]
This works for me and @siteshwar: `meson test --wrapper valgrind`. You can append a specific test name such as `builtins`.

> I do hope though that it helps me spot the SIGABRT and SIGSEV under macOS 10.12.6.

Note that termination due to SIGABRT is because the `abort()` function was called. In my experience it is rare for valgrind to increase the frequency of such failures. They typically represent "can't happen" situations that tend to be readily reproducible once you know the trigger condition. Valgrind does, however, greatly increase the likelihood of detecting problems that typically result in a SIGSEGV failure.

> In other words, now is not a good time to incorporated it.

Agreed. Once we have clean, no failure, test runs on the major platforms we're currently running tests on (macOS, FreeBSD, various Linux) we'll enable valgrind at least on Travis. Thanks for thinking about these things.

**Comment by krader1961** [comment]
Also, you'll need to increase the allowed test duration. The last time I did this I found I needed a very large multiplier and had to use `-t 50` on my slowest VM; i.e., `meson test -t 50 --wrapper valgrind`.

**Comment by catull** [comment]
Which version of meson do you use ?
Mine is 0.44.

None of these work for me:

`meson --wrapper valgrind`

```sh
usage: meson [-h] [--prefix PREFIX] [--libdir LIBDIR]
             [--libexecdir LIBEXECDIR] [--bindir BINDIR] [--sbindir SBINDIR]
             [--includedir INCLUDEDIR] [--datadir DATADIR] [--mandir MANDIR]
             [--infodir INFODIR] [--localedir LOCALEDIR]
             [--sysconfdir SYSCONFDIR] [--localstatedir LOCALSTATEDIR]
             [--sharedstatedir SHAREDSTATEDIR]
             [--backend {ninja,vs,vs2010,vs2015,vs2017,xcode}]
             [--buildtype {plain,debug,debugoptimized,release,minsize}]
             [--strip] [--unity {on,off,subprojects}] [--werror]
             [--layout {mirror,flat}] [--default-library {shared,static}]
             [--warnlevel {1,2,3}] [--stdsplit] [--errorlogs]
             [--cross-file CROSS_FILE] [-D option] [-v]
             [--wrap-mode {WrapMode.default,WrapMode.nofallback,WrapMode.nodownload}]
             [directories [directories ...]]
meson: error: unrecognized arguments: --wrapper
```

**Comment by krader1961** [comment]
Closing since I think this has been answered. I've also created a *CONTRIBUTING.md* document (currently out for review) that documents how to run the tests under Valgrind.
